skpro/datatypes/_convert.py

```diff
@@ -0,0 +1,310 @@
+        valid mtype strings, with explanation, are in datatypes.MTYPE_REGISTER
+    to_type : str - the type to convert "obj" to, a valid mtype string
+        valid mtype strings, with explanation, are in datatypes.MTYPE_REGISTER
+    as_scitype : str, optional - name of scitype the object "obj" is considered as
+        default = inferred from from_type
+        valid scitype strings, with explanation, are in datatypes.SCITYPE_REGISTER
+    store : optional, reference of storage for lossy conversions, default=None (no ref)
+        is updated by side effect if not None and store_behaviour="reset" or "update"
+    store_behaviour : str, optional, one of None (default), "reset", "freeze", "update"
+        "reset" - store is emptied and then updated from conversion
+        "freeze" - store is read-only, may be read/used by conversion but not changed
+        "update" - store is updated from conversion and retains previous contents
+        None - automatic: "update" if store is empty and not None; "freeze", otherwise
+
+    Returns
+    -------
+    converted_obj : to_type - object obj converted to to_type
+                    if obj was None, returns None
+
+    Raises
+    ------
+    KeyError if conversion is not implemented
+    TypeError or ValueError if inputs do not match specification
+    """
+    if obj is None:
+        return None
+
+    # input type checks
+    if not isinstance(to_type, str):
+        raise TypeError("to_type must be a str")
+    if not isinstance(from_type, str):
+        raise TypeError("from_type must be a str")
+    if as_scitype is None:
+        as_scitype = mtype_to_scitype(to_type)
+    elif not isinstance(as_scitype, str):
+        raise TypeError("as_scitype must be str or None")
+    if store is not None and not isinstance(store, dict):
+        raise TypeError("store must be a dict or None")
+    if store_behaviour not in [None, "reset", "freeze", "update"]:
+        raise ValueError(
+            'store_behaviour must be one of "reset", "freeze", "update", or None'
+        )
+    if store_behaviour is None and store == {}:
+        store_behaviour = "update"
+    if store_behaviour is None and store != {}:
+        store_behaviour = "freeze"
+
+    key = (from_type, to_type, as_scitype)
+
+    if key not in convert_dict.keys():
+        raise NotImplementedError(
+            "no conversion defined from type " + str(from_type) + " to " + str(to_type)
+        )
+
+    if store_behaviour == "freeze":
+        store = deepcopy(store)
+    elif store_behaviour == "reset":
+        # note: this is a side effect on store
+        store.clear()
+    elif store_behaviour == "update":
+        # store is passed to convert_obj by reference, unchanged
+        # this "elif" is here for clarity, to cover all three values
+        pass
+    else:
+        raise RuntimeError(
+            "bug: unrechable condition error, store_behaviour has unexpected value"
+        )
```

Comment:
nit: seems like a `ValueError` rather than runtime since it could be caught statically as well.

Reply:
yes, ok, changed